options/rtdl/generic/linker.cpp

```diff
@@ -1107,31 +1107,40 @@ frg::optional<ObjectSymbol> Scope::resolveSymbol(ObjectSymbol r, ResolveFlags fl
 
 Loader::Loader(Scope *scope, bool is_initial_link, uint64_t rts)
 : _globalScope{scope}, _isInitialLink{is_initial_link}, _linkRts{rts},
-		_linkSet{frg::hash<SharedObject *>{}, getAllocator()},
 		_linkBfs{getAllocator()}, _initQueue{getAllocator()} { }
 
-// TODO: Use an explicit vector to reduce stack usage to O(1)?
 void Loader::submitObject(SharedObject *object) {
-	if(_linkSet.get(object))
-		return;
-
-	// At this point the object is loaded and we can fill in its debug struct,
-	// the linked list fields will be filled later.
-	object->linkMap.base = object->baseAddress;
-	object->linkMap.name = object->path.data();
-	object->linkMap.dynv = object->dynamic;
+	__ensure(_linkBfs.size() == 0);
 
-	_linkSet.insert(object, Token{});
+	struct Token {};
+	using Set = frg::hash_map<SharedObject *, Token,
+			frg::hash<SharedObject *>, MemoryAllocator>;
+	Set set{frg::hash<SharedObject *>{}, getAllocator()};
```

Comment:
Why is this set moved out of the `Loader` object? This adds the (otherwise unnecessary?) restriction that `submitObject` must only be called once.

Reply:
I guess that would also work.

Reply:
Ok, done.